http://bugs.gentoo.org/353814
http://sources.redhat.com/bugzilla/show_bug.cgi?id=12489

From 996cf2ef07277805f119c8b47be92334c907ce8f Mon Sep 17 00:00:00 2001
From: Andreas Schwab <schwab@redhat.com>
Date: Sun, 20 Feb 2011 07:24:56 -0500
Subject: [PATCH] Move setting variable in relro data earlier in ld.so.

(cherry picked from commit e23fe25b33324a9ea992276c1a4f04127bf9ba4b)
---
 ChangeLog  |    8 ++++++++
 NEWS       |    2 +-
 elf/rtld.c |   11 +++++++----
 3 files changed, 16 insertions(+), 5 deletions(-)

2010-09-28  Andreas Schwab  <schwab@redhat.com>
	    Ulrich Drepper  <drepper@gmail.com>

	[BZ #12489]
	* elf/rtld.c (dl_main): Move setting of GLRO(dl_init_all_dirs)
	before performing relro protection.  At old place add assertion
	to make sure nothing changed.

diff --git a/elf/rtld.c b/elf/rtld.c
index 8510380..174954b 100644
--- a/elf/rtld.c
+++ b/elf/rtld.c
@@ -2179,6 +2179,10 @@ ERROR: ld.so: object '%s' cannot be loaded as audit interface: %s; ignored.\n",
      we need it in the memory handling later.  */
   GLRO(dl_initial_searchlist) = *GL(dl_ns)[LM_ID_BASE]._ns_main_searchlist;
 
+  /* Remember the last search directory added at startup, now that
+     malloc will no longer be the one from dl-minimal.c.  */
+  GLRO(dl_init_all_dirs) = GL(dl_all_dirs);
+
   if (prelinked)
     {
       if (main_map->l_info [ADDRIDX (DT_GNU_CONFLICT)] != NULL)
@@ -2298,9 +2302,8 @@ ERROR: ld.so: object '%s' cannot be loaded as audit interface: %s; ignored.\n",
 			  lossage);
     }
 
-  /* Remember the last search directory added at startup, now that
-     malloc will no longer be the one from dl-minimal.c.  */
-  GLRO(dl_init_all_dirs) = GL(dl_all_dirs);
+  /* Make sure no new search directories have been added.  */
+  assert (GLRO(dl_init_all_dirs) == GL(dl_all_dirs));
 
   if (! prelinked && rtld_multiple_ref)
     {
-- 
1.7.5.rc3

